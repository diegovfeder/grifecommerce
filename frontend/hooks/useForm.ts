import { useEffect, useState } from 'react';

interface IEvent {
	target: {
		files?: any;
		value?: any;
		name?: any;
		type?: any;
	};
}

// FIXME: What happens when I need another page to use this hook?
// Is there a way to define types Dynamically? Pass as props maybe?
interface IFormInput {
	name: string;
	description: string;
	price: number | undefined;
	image?: any;
}

export default function useForm(initial: IFormInput) {
	const [inputs, setInputs] = useState(initial);
	const initialValues = Object.values(initial).join('');

	useEffect(() => {
		setInputs(initial);
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [initialValues]);

	function handleChange(e: IEvent) {
		let { value, name, type } = e.target;
		if (type === 'number') {
			value = parseInt(value);
		}
		if (type === 'file') {
			[value] = e.target.files;
		}
		setInputs({
			...inputs,
			[name]: value,
		});
	}

	function resetForm() {
		setInputs(initial);
	}

	function clearForm() {
		const blankState = Object.fromEntries(
			Object.entries(inputs).map(([key, _]) => [key, '']),
		);
		setInputs(blankState);
	}

	return {
		inputs,
		handleChange,
		resetForm,
		clearForm,
	};
}

// const initialValues = Object.values(initial).join('');
// useEffect(() => {
// 	// This function runs when the things we are watching change
// 	setInputs(initial);
// }, [initial]);

// {
//   name: 'wes',
//   description: 'nice shoes',
//   price: 1000
// }
