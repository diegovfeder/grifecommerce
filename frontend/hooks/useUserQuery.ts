import { gql, useQuery } from '@apollo/client';

export const CURRENT_USER_QUERY = gql`
	query CurrentUserQuery {
		authenticatedItem {
			... on User {
				id
				email
				name
				cart {
					id
					quantity
					product {
						id
						price
						name
						description
						photo {
							image {
								publicUrlTransformed
							}
						}
					}
				}
			}
		}
	}
`;

const useUserQuery = () => {
	const { data } = useQuery(CURRENT_USER_QUERY);
	return data?.authenticatedItem;
};

export default useUserQuery;
