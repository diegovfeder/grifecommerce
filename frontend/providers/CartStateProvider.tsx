import React, {
	createContext,
	useState,
	Dispatch,
	SetStateAction,
} from 'react';

interface LocalCartStateProps {
	cartOpen: boolean;
	setCartOpen: Dispatch<SetStateAction<boolean>>;
	closeCart: () => void;
	openCart: () => void;
	toggleCart: () => void;
}

export const LocalCartStateContext = createContext<LocalCartStateProps>({
	cartOpen: false,
	setCartOpen: () => {},
	closeCart: () => {},
	openCart: () => {},
	toggleCart: () => {},
});
const LocalCartStateProvider = LocalCartStateContext.Provider;

const CartStateProvider = ({ children }: { children: React.ReactNode }) => {
	const [cartOpen, setCartOpen] = useState(false);

	const closeCart = () => {
		setCartOpen(false);
	};

	const openCart = () => {
		setCartOpen(true);
	};

	const toggleCart = () => {
		setCartOpen(!cartOpen);
	};

	return (
		<LocalCartStateProvider
			value={{ cartOpen, setCartOpen, closeCart, openCart, toggleCart }}
		>
			{children}
		</LocalCartStateProvider>
	);
};

export default CartStateProvider;
