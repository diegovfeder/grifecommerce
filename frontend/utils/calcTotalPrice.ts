import { CartItemProps } from '../@types/commonTypes';

export default function calcTotalPrice(cart: CartItemProps[]) {
	return cart.reduce((tally, cartItem) => {
		if (!cartItem.product) return tally;
		return tally + cartItem.quantity * cartItem.product.price;
	}, 0);
}
