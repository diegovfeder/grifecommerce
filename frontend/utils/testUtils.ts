// eslint-disable-next-line import/no-extraneous-dependencies
import casual from 'casual';
import { PRODUCTS_COUNT_QUERY } from '../gql/queries';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
	// __typename: 'Item',
	id: '718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d',
	name: 'GRIFE Sample Pack - (Yellow)',
	description:
		'This our first sample pack for summer beats in a sunny day, when all is yellow :)',
	price: 5000,
	// user: null,
	photo: {
		// FIXME:
		// id: 'abc123',
		// altText: 'dogs are best',
		image: {
			publicUrlTransformed: '/samplepack_yellow.jpg',
		},
	},
});

const fakeUser = (rest?: any) => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
	...rest,
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: {
		image: `${casual.word}.jpg`,
	},
	name: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words(),
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2022-12-11T20:16:13.797Z',
	user: fakeUser(),
});

const fakeCartItem = (rest?: any) => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	product: fakeItem(),
	user: fakeUser(),
	...rest,
});

// FIXME: Properly type this class
// Fake LocalStorage
class LocalStorageMock {
	store: { [key: string]: any } = {};
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key: string) {
		return this.store[key] || null;
	}

	setItem(key: string, value: string) {
		this.store[key] = value.toString();
	}

	removeItem(key: string) {
		delete this.store[key];
	}
}

function makePaginationMocksFor(length: number) {
	return [
		{
			request: { query: PRODUCTS_COUNT_QUERY },
			result: {
				data: {
					_allProductsMeta: {
						count: length,
					},
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							count: length,
							__typename: 'count',
						},
					},
				},
			},
		},
	];
}

export {
	makePaginationMocksFor,
	LocalStorageMock,
	fakeItem,
	fakeUser,
	fakeCartItem,
	fakeOrder,
	fakeOrderItem,
};
