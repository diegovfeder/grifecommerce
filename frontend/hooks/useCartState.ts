import { useContext } from 'react';
import { LocalCartStateContext } from '../providers/CartStateProvider';

const useCartState = () => {
	const cartState = useContext(LocalCartStateContext);
	return cartState;
};

export default useCartState;
