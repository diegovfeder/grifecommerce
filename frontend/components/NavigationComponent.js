import Link from 'next/link';
import SignOut from './SignOut';
import CartCount from './CartCount';
import StyledNavigation from './styles/StyledNavigation';
import useCartState  from '../hooks/useCartState';
import useUserQuery  from '../hooks/useUserQuery';

const NavigationComponent = () => {
	const user = useUserQuery();
	const { openCart } = useCartState();
	return (
		<StyledNavigation>
			<Link href="/products">Products</Link>
			{user && (
				<>
					<Link href="/sell">Sell</Link>
					<Link href="/orders">Orders</Link>
					<Link href="/account">Account</Link>
					<SignOut />
					<button type="button" onClick={openCart}>
						My Cart
						<CartCount
							count={user.cart.reduce(
								(tally, cartItem) =>
									tally + (cartItem.product ? cartItem.quantity : 0),
								0
							)}
						/>
					</button>
				</>
			)}
			{!user && (
				<>
					<Link href="/signin">Sign In</Link>
				</>
			)}
		</StyledNavigation>
	);
}

export default NavigationComponent;
