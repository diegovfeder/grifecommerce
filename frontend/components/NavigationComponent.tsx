import Link from 'next/link';

import useCartState from '../hooks/useCartState';
import useUserQuery from '../hooks/useUserQuery';
import { CartItemProps } from '../types/commonTypes';
import StyledNavigation from './styles/StyledNavigation';
import CartCount from './CartCount';
import SignOut from './SignOut';

const NavigationComponent = () => {
	const user = useUserQuery();
	const { openCart } = useCartState();
	return (
		<StyledNavigation>
			<Link href="/products/1">PRODUCTS</Link>
			{user ? (
				<>
					<Link href="/sell">SELL</Link>
					<Link href="/orders">ORDERS</Link>
					<Link href="/account">ACCOUNT</Link>
					<SignOut />
					<button type="button" onClick={openCart}>
						MY CART
						<CartCount
							count={user.cart.reduce(
								(tally: number, cartItem: CartItemProps) =>
									tally + (cartItem.product ? cartItem.quantity : 0),
								0,
							)}
						/>
					</button>
				</>
			) : (
				<>
					<Link href="/signin">SIGN IN</Link>
				</>
			)}
		</StyledNavigation>
	);
};

export default NavigationComponent;
