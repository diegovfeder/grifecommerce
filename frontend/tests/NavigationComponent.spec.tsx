import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import NavigationComponent from '../components/NavigationComponent';
import { CURRENT_USER_QUERY } from '../hooks/useUserQuery';
import { fakeUser, fakeCartItem } from '../utils/testUtils';
import CartStateProvider from '../providers/CartStateProvider';

const notSignedInMocks = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: { data: { authenticatedItem: null } },
	},
];

const signedInMocks = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: { data: { authenticatedItem: fakeUser() } },
	},
];

const signedInMocksWithCartItems = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: {
			data: {
				authenticatedItem: fakeUser({
					cart: [fakeCartItem()],
				}),
			},
		},
	},
];

describe('<NavigationComponent/>', () => {
	it('renders navigation when signed out', () => {
		const { container, debug } = render(
			<CartStateProvider>
				<MockedProvider mocks={notSignedInMocks}>
					<NavigationComponent />
				</MockedProvider>
			</CartStateProvider>,
		);
		expect(container).toHaveTextContent(/sign in/i);
		expect(container).toMatchSnapshot();
		const link = screen.getByText(/sign in/i);
		expect(link).toHaveAttribute('href', '/signin');
		const productsLink = screen.getByText(/products/i);
		expect(productsLink).toBeInTheDocument();
		expect(productsLink).toHaveAttribute('href', '/products/1');
	});

	it('renders navigation when signed in', async () => {
		const { container, debug } = render(
			<CartStateProvider>
				<MockedProvider mocks={signedInMocks}>
					<NavigationComponent />
				</MockedProvider>
			</CartStateProvider>,
		);
		await screen.findByText(/account/i);
		expect(container).toMatchSnapshot();
		expect(container).toHaveTextContent(/sign out/i);
		expect(container).toHaveTextContent(/my cart/i);
	});

	it('renders the amount of items in the cart', async () => {
		const { container, debug } = render(
			<CartStateProvider>
				<MockedProvider mocks={signedInMocksWithCartItems}>
					<NavigationComponent />
				</MockedProvider>
			</CartStateProvider>,
		);
		await screen.findByText(/account/i);
		expect(screen.getByText('3')).toBeInTheDocument();
	});
});
