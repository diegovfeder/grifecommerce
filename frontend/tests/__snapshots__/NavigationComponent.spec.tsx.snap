// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationComponent/> renders navigation when signed in 1`] = `
<div>
  <ul
    class="StyledNavigation-sc-hq3jod-0 efkomq"
  >
    <a
      href="/products/1"
    >
      PRODUCTS
    </a>
    <a
      href="/sell"
    >
      SELL
    </a>
    <a
      href="/orders"
    >
      ORDERS
    </a>
    <a
      href="/account"
    >
      ACCOUNT
    </a>
    <button
      type="button"
    >
      SIGN OUT
    </button>
    <button
      type="button"
    >
      MY CART
      <span
        class="CartCount__StyledAnimation-sc-181jath-1 gKoXcI"
      >
        <div>
          <div
            class="CartCount__StyledDot-sc-181jath-0 cfzsbD count"
          >
            0
          </div>
        </div>
      </span>
    </button>
  </ul>
</div>
`;

exports[`<NavigationComponent/> renders navigation when signed out 1`] = `
<div>
  <ul
    class="StyledNavigation-sc-hq3jod-0 efkomq"
  >
    <a
      href="/products/1"
    >
      PRODUCTS
    </a>
    <a
      href="/signin"
    >
      SIGN IN
    </a>
  </ul>
</div>
`;
